scipy/sparse/construct.py

```diff
@@ -667,16 +667,33 @@ def block_diag(mats, format=None, dtype=None):
            [0, 0, 0, 7]])
 
     """
-    nmat = len(mats)
-    rows = []
-    for ia, a in enumerate(mats):
-        row = [None]*nmat
+    row = []
+    col = []
+    data = []
+    r_idx = 0
+    c_idx = 0
+    for a in mats:
+        if isinstance(a, (list, numbers.Number)):
+            a = coo_matrix(a)
+        nrows, ncols = a.shape
         if issparse(a):
-            row[ia] = a
+            a = a.tocoo()
+            row.append(a.row + r_idx)
+            col.append(a.col + c_idx)
+            data.append(a.data)
         else:
-            row[ia] = coo_matrix(a)
-        rows.append(row)
-    return bmat(rows, format=format, dtype=dtype)
+            a_row, a_col = np.divmod(np.arange(nrows*ncols), ncols)
```

Comment:
These can be constructed directly, avoiding the divisions:

```
a_row = np.repeat(np.arange(r_idx, r_idx + nrows), ncols)
a_col = np.tile(np.arange(c_idx, c_idx + ncols), nrows)
```

Reply:
Those changed are clearer but they make the code slower; see [my previous comment](https://github.com/scipy/scipy/pull/12356#issuecomment-652629014).

Reply:
Up @perimosocordiae.